src/chainrules.jl

```diff
@@ -0,0 +1,158 @@
+## Forward Rules
+
+function ChainRulesCore.frule((_, Δx, Δy), d::Distances.Euclidean, x, y)
+    Δ = x - y
+    D = sqrt(sum(abs2, Δ))
+    if iszero(D)
+        return D, Zero()
+    else
+        Δ ./= D
+        return D, dot(Δ, Δx) - dot(Δ, Δy)
+    end
+end
+
+## Reverse Rules Delta
+
+function ChainRulesCore.rrule(::typeof(Distances.evaluate), s::Delta, x::AbstractVector, y::AbstractVector)
+    d = evaluate(s, x, y)
+    function evaluate_pullback(::Any)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return d, evaluate_pullback
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix; dims=2
+)
+    P = Distances.pairwise(d, X, Y; dims=dims)
+    function pairwise_pullback(::Any)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix; dims=2)
+    P = Distances.pairwise(d, X; dims=dims)
+    function pairwise_pullback(::Any)
+        return NO_FIELDS, Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.colwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix)
+    C = Distances.colwise(d, X, Y)
+    function colwise_pullback(::AbstractVector)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return C, colwise_pullback
+end
+
+## Reverse Rules DotProduct
+function ChainRulesCore.rrule(
+    ::typeof(Distances.evaluate), s::DotProduct, x::AbstractVector, y::AbstractVector
+)
```

Comment:
Same here, maybe it would be better to define an adjoint for
```suggestion
function ChainRulesCore.rrule(
    s::DotProduct, x::AbstractVector, y::AbstractVector
)
```
(if we overload `DotProduct`)